lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AirCompressorCodec.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.hive.formats.avro;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.compress.Compressor;
+import io.airlift.compress.Decompressor;
+import io.airlift.compress.snappy.SnappyCompressor;
+import io.airlift.compress.snappy.SnappyDecompressor;
+import io.airlift.compress.zstd.ZstdCompressor;
+import io.airlift.compress.zstd.ZstdDecompressor;
+import org.apache.avro.file.Codec;
+import org.apache.avro.file.CodecFactory;
+import org.apache.avro.file.DataFileConstants;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class AirCompressorCodec
+        extends Codec
+{
+    @VisibleForTesting
+    public static final Supplier<AirCompressorCodec> SNAPPY_CODEC_SUPPLIER = () -> new AirCompressorCodec(DataFileConstants.SNAPPY_CODEC, new SnappyCompressor(), new SnappyDecompressor(), bytes -> SnappyDecompressor.getUncompressedLength(bytes, 0), true);
+
+    public static final CodecFactory SNAPPY_CODEC_FACTORY = new CodecFactory()
+    {
+        @Override
+        protected Codec createInstance()
+        {
+            return SNAPPY_CODEC_SUPPLIER.get();
+        }
+    };
+
+    @VisibleForTesting
+    public static final Supplier<AirCompressorCodec> ZSTD_CODEC_SUPPLIER = () -> new AirCompressorCodec(DataFileConstants.ZSTANDARD_CODEC, new ZstdCompressor(), new ZstdDecompressor(), bytes -> {
+        int decompressedSize = toIntExact(ZstdDecompressor.getDecompressedSize(bytes, 0, bytes.length));
+        if (decompressedSize < 0) {
+            return bytes.length * 2;
+        }
+        return decompressedSize;
+    }, false);
+
+    public static final CodecFactory ZSTD_CODEC_FACTORY = new CodecFactory()
+    {
+        @Override
+        protected Codec createInstance()
+        {
+            return ZSTD_CODEC_SUPPLIER.get();
+        }
+    };
+
+    private final String name;
+    private final Compressor compressor;
+    private final Decompressor decompressor;
+    private final Function<byte[], Integer> decompressionSizeGeneration;
+    private final boolean includeCrc32;
+    private final CRC32 crc32 = new CRC32();
```

Comment:
I'm thinking it would be better to have `SnappyAirCompressorCodec` and `ZstdAirCompressorCodec`, since there are only two instances, then we don't need this flag. There will be a bit of duplication, but that seems cleaner than a special purpose flag here.